client/src/components/Header.test.js

```diff
@@ -0,0 +1,231 @@
+  });
+
+  it("Should render cart count correctly", () => {
+    render(
+      <MemoryRouter>
+        <Header />
+      </MemoryRouter>
+    );
+
+    expect(screen.getByText("2")).toBeInTheDocument();
+  });
+
+  describe("When user is not authenticated", () => {
+    beforeEach(() => {
+      jest.clearAllMocks();
+    });
+
+    it("Should render link to register page", () => {
+      render(
+        <MemoryRouter>
+          <Header />
+        </MemoryRouter>
+      );
+      const registerLink = screen.getByRole("link", { name: "Register" });
+      expect(registerLink).toBeInTheDocument();
+      expect(registerLink).toHaveAttribute("href", "/register");
+    });
+
+    it("Should render link to login page", () => {
+      render(
+        <MemoryRouter>
+          <Header />
+        </MemoryRouter>
+      );
+      const loginLink = screen.getByRole("link", { name: "Login" });
+      expect(loginLink).toBeInTheDocument();
+      expect(loginLink).toHaveAttribute("href", "/login");
+    });
+  });
+
+  describe("When user is authenticated", () => {
+    const mockSetAuth = jest.fn();
+
+    beforeEach(() => {
+      jest.clearAllMocks();
+      useAuth.mockImplementation(() => [
+        { user: { name: "user", role: 0 } },
+        mockSetAuth,
+      ]);
+    });
+
+    it("Should render user's name", () => {
+      render(
+        <MemoryRouter>
+          <Header />
+        </MemoryRouter>
+      );
+      expect(screen.getByText("user")).toBeInTheDocument();
+    });
+
+    describe("When user is not an admin", () => {
+      beforeEach(() => {
+        jest.clearAllMocks();
+        useAuth.mockImplementation(() => [
+          { user: { name: "user", role: 0 } },
+          mockSetAuth,
+        ]);
+      });
+
+      it("Should render link to user dashboard", async () => {
+        const user = userEvent.setup();
+
+        render(
+          <MemoryRouter>
+            <Header />
+          </MemoryRouter>
+        );
+
+        await user.click(screen.getByText("user"));
+        const dashboardLink = screen.getByRole("link", { name: "Dashboard" });
+        expect(dashboardLink).toBeInTheDocument();
+        expect(dashboardLink).toHaveAttribute("href", "/dashboard/user");
+      });
+    });
+
+    it("Should render logout button", () => {
+      render(
+        <MemoryRouter>
+          <Header />
+        </MemoryRouter>
+      );
+      const logoutButton = screen.getByRole("link", { name: "Logout" });
+      expect(logoutButton).toBeInTheDocument();
+      expect(logoutButton).toHaveAttribute("href", "/login");
+    });
+
+    it("Should log user out when logout button is clicked", async () => {
+      const user = userEvent.setup();
+
+      render(
+        <MemoryRouter>
+          <Header />
+        </MemoryRouter>
+      );
+
+      await user.click(screen.getByText("user"));
+      await user.click(screen.getByText("Logout"));
+      expect(localStorage.removeItem).toHaveBeenCalledWith("auth");
```

Comment:
I'm wondering if the cart should be cleared on logout as well? Although this means that cart items will not persist across sessions, unless we store it to the DB

Because currently, I can log in as User A, add items to my cart then logout. If I were to now log in as User B, my cart contains all User A's cart items and I can modify User A's cart as User B. (the cart is shared since it's stored and retrieved from local storage)